### ISSUE TITLE
Small glitches on heatmaps

### ISSUE THREAD

**Issue by markomanninen** [matplotlib/matplotlib#9831]
### Bug report

I found visible glitches on heatmaps on otherwise uniform data. Is it possible to outcome these by some setting and what causes them?

**Code for reproduction**

I'm using Seaborn API for the heatmap, but according to Seaborn developer it straightly uses matplotlib. 

https://github.com/mwaskom/seaborn/issues/1331

At the moment I have no exact matplotlib code for the reproduction, but I have Jupyter notebook document attached to see the whole code and several examples for the glitch.

In Seaborn API I use simply sns.heatmap and provide two dimentional data by coeftt function:

```python
sns.heatmap(coeftt(range(1, 21), max))
```

**Actual outcome**

![maxcoeffs21](https://user-images.githubusercontent.com/42345/33111635-cfdd8a3a-cf57-11e7-83be-d16ed324b1a5.png)

There you can see couple of small squares in rows 3 and 4, columns 9 and 19.

Uniformness of the data should be visible in this screenshot, also reacheable on nb doc:

![maxcoeffsd](https://user-images.githubusercontent.com/42345/33111718-236fcbcc-cf58-11e7-886b-4c322f4d2811.png)

**Matplotlib version**

  * Operating system: Windows 10
  * Matplotlib version: 2.1.0
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://ipykernel.pylab.backend_inline
  * Python version: 3.6.2
  * Jupyter version (if applicable): packaged by conda-forge | (default, Jul 23 2017, 22:58:45) [MSC v.1900 64 bit (AMD64)]
  * Other libraries: Seaborn 0.8.1

Matplotlib was installed by anaconda, default channel, came together with packages in Seaborn.

[MaxCoefficientonPolynomial.zip](https://github.com/matplotlib/matplotlib/files/1494722/MaxCoefficientonPolynomial.zip)

**Comment by markomanninen** [comment]
Just an update, I got better results using purely matplotlib:

```
import matplotlib.pyplot as plt
import numpy as np
plt.rcParams["axes.grid"] = False
plt.imshow(coeftt(range(1, 21), max))
plt.show()
```

![maxcoeffs](https://user-images.githubusercontent.com/42345/33116040-e4028d6c-cf6b-11e7-981a-c6958eabb5c1.png)


so I guess I need to turn back to Seaborn support.

**Comment by tacaswell** [comment]
Are these glitches dependent on the size and dpi of the figure you are saving?

**Comment by mwaskom** [comment]
Seems dependent on y axis inversion:

```
f, ax = plt.subplots()
ax.pcolormesh(coeftt(l, max))
```

![image](https://user-images.githubusercontent.com/315810/33221605-c9e81992-d11e-11e7-8a13-a155f2f39970.png)


```
f, ax = plt.subplots()
ax.pcolormesh(coeftt(l, max))
ax.invert_yaxis()
```

![image](https://user-images.githubusercontent.com/315810/33221604-c78c0b18-d11e-11e7-925a-b9db9ae668ce.png)

**Comment by jklymak** [comment]
On the macos backend I got this to show up once in a while.  If you change the size of the figure, it goes away and comes back.  Of course my dpi is 200, so it'd be harder to see than on a 120 dpi display.  

```
import numpy as np
import matplotlib.pyplot as plt

def _coef(n):
    # 216 -> poly1d([2, 1, 6])
    return np.poly1d(list(map(int, str(n))))

def coef(i, j, cb):
    return cb(_coef(i) * _coef(j)) + 1

def coeft(l, cb):
    for i in l:
        mm = []
        for j in l:
            mm.append(coef(i, j, cb))
        yield mm

def coeftt(l, cb):
    m = []
    for i in l:
        mm = []
        for j in l:
            mm.append(coef(i, j, cb))
        m.append(mm)
    return m

xx = coeftt(range(1, 40), max)

fig, ax = plt.subplots()
ax.pcolormesh(xx)
plt.show()

fig, ax = plt.subplots()
ax.pcolormesh(xx)
ax.invert_yaxis()
plt.show()
```

**Comment by tacaswell** [comment]
I can not reproduce this, but I only played around with it interactively.

**Comment by QuLogic** [comment]
I could reproduce in 2.0.0 with @jklymak's example, though I had to expand the window a little bit wider.

**Comment by QuLogic** [comment]
This was fixed by #16090.
